src/AssociatedJobInfoClient.cs

```diff
@@ -0,0 +1,124 @@
+using Microsoft.DotNet.HelixPoolProvider.Models;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.HelixPoolProvider
+{
+    public class AssociatedJobInfo
+    {
+        public string BuildSourceBranch { get; }
+        public string SystemPullRequestTargetBranch { get; }
+
+        public static readonly AssociatedJobInfo Empty
+            = new AssociatedJobInfo("", "");
+
+        public AssociatedJobInfo(
+            string buildSourceBranch,
+            string systemPullRequestTargetBranch)
+        {
+            BuildSourceBranch = buildSourceBranch;
+            SystemPullRequestTargetBranch = systemPullRequestTargetBranch;
+        }
+    }
+
+    public class AssociatedJobInfoClient
+    {
+        private readonly IHttpClientFactory _httpClientFactory;
+        private readonly ILogger<AssociatedJobInfoClient> _logger;
+
+        public AssociatedJobInfoClient(
+            IHttpClientFactory httpClientFactory,
+            ILogger<AssociatedJobInfoClient> logger)
+        {
+            _httpClientFactory = httpClientFactory;
+            _logger = logger;
+        }
+
+        public async Task<AssociatedJobInfo> TryGetAssociatedJobInfo(
+            string getAssociatedJobUrl,
+            string authenticationToken)
+        {
+            try
+            {
+                var agentRequestJob = await TryGetAssociatedAgentRequestJob(
+                    getAssociatedJobUrl,
+                    authenticationToken);
+
+                if (agentRequestJob != null)
+                {
+                    return ParseAssociatedJobInfo(agentRequestJob);
+                }
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError(
+                    exception,
+                    "Unable to get associated job info from {getAssociatedJobUrl} because of an exception",
+                    getAssociatedJobUrl);
+            }
+
+            return AssociatedJobInfo.Empty;
+        }
+
+        private async Task<AgentRequestJob> TryGetAssociatedAgentRequestJob(
+            string getAssociatedJobUrl,
+            string authenticationToken)
+        {
+            if (!string.IsNullOrEmpty(getAssociatedJobUrl))
+            {
+                _logger.LogInformation("Getting associated job from {AssociatedJobUrl}", getAssociatedJobUrl);
+
+                var httpClient = _httpClientFactory.CreateClient();
+                var message = new HttpRequestMessage(HttpMethod.Get, getAssociatedJobUrl);
+                message.Headers.Authorization = new AuthenticationHeaderValue("Bearer", authenticationToken);
+                var response = await httpClient.SendAsync(message);
```

Comment:
This needs a relatively short time out.  If we don't respond by 30 seconds (and it gets worse the longer we take) the way pool providers work means this slows down all agent pools in the entire project badly.

Reply:
I have set the timeout to 10 seconds.